server/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -847,11 +848,34 @@ public void forceMerge(boolean flush) throws IOException {
      */
     public abstract IndexCommitRef acquireSafeIndexCommit() throws EngineException;
 
+    /**
+     * If the specified throwable contains a fatal error in the throwable graph, such a fatal error will be thrown. Callers should ensure
+     * that there are no catch statements that would catch an error in the stack as the fatal error here should go uncaught and be handled
+     * by the uncaught exception handler that we install during bootstrap. If the specified throwable dies indeed contain a fatal error, the
+     * specified message will attempt to be logged before throwing the fatal error. If the specified throwable does not contain a fatal
+     * error, this method is a no-op.
+     *
+     * @param maybeMessage the message to maybe log
+     * @param maybeFatal   the throwable that maybe contains a fatal error
+     */
+    @SuppressWarnings("finally")
+    private void maybeDie(final String maybeMessage, final Throwable maybeFatal) {
+        final Optional<Error> maybeError = ExceptionsHelper.maybeError(maybeFatal, logger);
```

Comment:
As long as you're using Optional here, I think this could be

```java
ExceptionsHelper.maybeError(maybeFatal, logger).ifPresent(e -> {
    try {
         logger.error(maybeMessage, e);
    } finally {
        throw e;
    }});
```

Up to you if you want though

Reply:
Thanks.